Document MSRV policy (#881)
mmtk/mmtk-core: README.md

MMTk is also tested with the VM bindings we are maintaining by running standard test/benchmark suites for the VMs.
For details, please refer to each VM binding repository.

## Supported Rust versions

> As a research project, MMTk is quite aggressive in following the latest releases of the Rust

Author: Fixed.

> toolchain. The minimum supported Rust version (MSRV) is "N-1", that is, one minor version before the

Author: I think that's also OK.  My understanding is,

1.  We always guarantee that "N-1" is supported.
2.  But the actual `rust-version` field in `Cargo.toml` *may* sometimes be older than `N-1` if we didn't make use of any newer features during the last 6 weeks of development, and no dependencies require a newer `rust-version`.
3.  If, during our development, we made a PR that uses a newer feature introduced in `N-m` where `M>=1`, we have a good reason to bump `rust-version` to `N-m`.
4.  The CI shall test with both the latest stable version and the `rust-version` in `Cargo.toml`.

That still sounds good, but I still prefer bumping the MSRV eagerly (even for "no reason"), for the following reasons:
1.  If the MSRV is already bumped after a release, we don't have to deliberately bump it during development.
2.  If we cannot guarantee that `N-2` is supported, our users still cannot rely on `N-2` even if the `rust-version` in `Cargo.toml` is sometimes `N-2` or older.  That's because it may change in the next release.  Although the user may lock the version of `mmtk-core` to a version to satisfy their MSRV requirement, we should not encouraged this (at least for now) because MMTk is being rapidly developed, and relying on an outdated MMTk now will make it difficult for the users to migrate to the latest MMTk in the future.

However, we may change the policy if MMTk reached a stable state.  By that time, we may prefer letting the MSRV "linger" at a version, like you suggested, if we do not make use of newer features.  That will allow the users to use a newer `mmtk-core` while still using an older Rust toolchain.  At that time, it should be reasonable for the user to lock the version of `mmtk-core` because `mmtk-core` is relatively stable.  But I don't think MMTk has reached such a state.

latest stable Rust release.  For example, if the latest Rust stable release is "1.61.2", then the
MSRV will be "1.60.0".

> MMTk has a six-week release cycle, roughly the same as Rust itself.  We bump the MSRV version

Author: >   Last time we talked with @caizixian on this. We agreed on bumping the version in rust-toolchain for every other release (he needs to update our dev and CI machines).

If updating the toolchains every 6 weeks is too much work, we can just say our MSRV is "N-2".  I am still OK with "N-2" or "N-3".  In the case of "N-2", MMTk-core will support three Rust stable versions, and each Rust version remain supported for 18 weeks.  If we update our toolchains for every even (or odd) Rust releases (that is, every 12 weeks), the existing installed toolchain will remain supported by the time we do our update.  In other word, on the day when an even Rust version (assume we update Rust at even versions) is released, both the currently installed toolchain and the newest toolchain will be supported, and we can do the update in the window of 6 weeks from that day until the next Rust release.  (Actually we may have up to 12 weeks because the MSRV of MMTk core is the "N-2" right after the *previous* MMTk core release.)

@caizixian What do you think of it? If "N-2" is still too much work, we can use "N-3" instead.

MMTk is also tested with the VM bindings we are maintaining by running standard test/benchmark suites for the VMs.
For details, please refer to each VM binding repository.

## Supported Rust versions

MMTk uses a pinned Rust version in the repository (recorded in the `rust-toolchain` file).  We run
our tests and benchmarks using the pinned Rust version. We recommend using the pinned Rust version
for development. We update the pinned Rust version between releases of mmtk-core to keep it close to
the latest Rust stable release. The release cycle of mmtk-core is six weeks, roughly the same as
Rust itself. 

> Our minimum support Rust version (MSRV) policy is "N-1".  That means we also ensure mmtk-core works

Reviewer: I think it's worth reminding that the N is NOT the current Rust release. N is the version recorded in `rust-toolchain`, which might lag behind the current stable Rust stable release.

properly with the Rust toolchain that is one minor version before the version specified in
`rust-toolchain`.  For example, if `rust-toolchain` contains "1.61.2", the MSRV will be guaranteed
to be no later than "1.60.0".  We may bump MSRV up to "N-1" when we need to make use of new Rust

> features or a dependency crate that needs a newer Rust. Users are encouraged to keep close to the

Reviewer: Perhaps we can clarify further.

```suggestion
features or a dependency crate that needs a newer Rust. In other words, we will not eagerly bump the MSRV, and the MSRV can lag behind "N-1" in practice. However, users shouldn't depend on this lag, and are encouraged to keep close to the
```

MMTk is also tested with the VM bindings we are maintaining by running standard test/benchmark suites for the VMs.
For details, please refer to each VM binding repository.

## Supported Rust versions

The MMTk team does most of the development and experiment work on a specific Rust version recorded
in the `rust-toolchain` file.  We update it between releases of mmtk-core to keep it close to the
latest Rust stable release.  The release cycle of mmtk-core is six weeks, roughly the same as Rust
itself. 

We also run CI tests against the latest stable Rust version for each pull request.  If it fails, we
will not bump the version in `rust-toolchain` until we fix the issues. 

Our minimum support Rust version (MSRV) policy is "N-1".  That means we also ensure mmtk-core works
properly with the Rust toolchain that is one minor version before the version specified in
`rust-toolchain`.  For example, if `rust-toolchain` contains "1.61.2", the MSRV will be "1.60.0".

> The `package.rust-version` field in `Cargo.toml` may sometimes contain an older version, but we may
> bump it to a newer version up to "N-1" whenever we need to make use of new Rust features or a
> dependency crate needs a newer Rust.  Therefore, users are encouraged to keep close to the latest
> Rust toolchain rather than relying on an outdated version of Rust.

Reviewer 2: ```suggestion
We may bump MSRV up to "N-1" when we need to make use of new Rust features or a
dependency crate that needs a newer Rust. Users are encouraged to keep close to the latest
Rust toolchain rather than relying on an outdated version of Rust.
```
